cpp/src/io/json/experimental/read_json.cpp

```diff
@@ -38,29 +39,47 @@ size_t sources_size(host_span<std::unique_ptr<datasource>> const sources,
   });
 }
 
-std::vector<uint8_t> ingest_raw_input(host_span<std::unique_ptr<datasource>> const& sources,
-                                      compression_type compression,
-                                      size_t range_offset,
-                                      size_t range_size)
+rmm::device_uvector<char> ingest_raw_input(host_span<std::unique_ptr<datasource>> const& sources,
+                                           compression_type compression,
+                                           size_t range_offset,
+                                           size_t range_size,
+                                           rmm::cuda_stream_view stream)
 {
   CUDF_FUNC_RANGE();
   // Iterate through the user defined sources and read the contents into the local buffer
   auto const total_source_size = sources_size(sources, range_offset, range_size);
-  auto buffer                  = std::vector<uint8_t>(total_source_size);
-
-  size_t bytes_read = 0;
-  for (const auto& source : sources) {
-    if (!source->is_empty()) {
-      auto data_size   = (range_size != 0) ? range_size : source->size();
-      auto destination = buffer.data() + bytes_read;
-      bytes_read += source->host_read(range_offset, data_size, destination);
-    }
-  }
 
   if (compression == compression_type::NONE) {
-    return buffer;
+    auto d_buffer     = rmm::device_uvector<char>(total_source_size, stream);
+    size_t bytes_read = 0;
+    for (const auto& source : sources) {
```

Comment:
Concatenating entire sources makes sense for json_lines.

This logic of concatenating multiple chunks (with range_offset) from different sources does not make sense for json, even for uncompressed. IIRC, I replicated this logic from another cudf code. May be we should rethink this logic again.

Reply:
I was wondering about some code in here :) For example, it's not clear to me what `byte_range` means with multiple sources.
What do you think about addressing this in a separate PR? This one should be focused on `host_read` -> `device_read` transition.
Is there an issue where we can discuss this further? I only opened https://github.com/rapidsai/cudf/issues/12299

Reply:
Let's fix it in another PR. This needs more discussion. byte range for different sources may be different, so we may request byte range for each source.
For now, it could throw error "Multiple-sources with byte range is not supported"  (non-zero range_size).